src/Database/TypeTrait.php

```diff
@@ -0,0 +1,159 @@
+        'time' => 'Cake\Datasource\Type\TimeType',
+        'timestamp' => 'Cake\Datasource\Type\DateTimeType',
+        'uuid' => 'Cake\Datasource\Type\UuidType',
+    ];
+
+
+    /**
+     * Contains a map of type object instances to be reused if needed
+     *
+     * @var array
+     */
+    protected static $_builtTypes = [];
+
+    /**
+     * Returns a Type object capable of converting a type identified by $name
+     *
+     * @param string $name type identifier
+     * @throws \InvalidArgumentException If type identifier is unknown
+     * @return Type
+     */
+    public static function build($name)
+    {
+        if (isset(static::$_builtTypes[$name])) {
+            return static::$_builtTypes[$name];
+        }
+        if (!isset(static::$_types[$name])) {
+            throw new InvalidArgumentException(sprintf('Unknown type "%s"', $name));
+        }
+        return static::$_builtTypes[$name] = new static::$_types[$name]($name);
+    }
+
+    /**
+     * Returns a Type object capable of converting a type identified by $name
+     *
+     * @param string $name The type identifier you want to set.
+     * @param \Cake\Database\Type $instance The type instance you want to set.
+     * @return void
+     */
+    public static function set($name, Type $instance)
+    {
+        static::$_builtTypes[$name] = $instance;
+    }
+
+    /**
+     * Registers a new type identifier and maps it to a fully namespaced classname,
+     * If called with no arguments it will return current types map array
+     * If $className is omitted it will return mapped class for $type
+     *
+     * @param string|array|null $type if string name of type to map, if array list of arrays to be mapped
+     * @param string|null $className The classname to register.
+     * @return array|string|null if $type is null then array with current map, if $className is null string
+     * configured class name for give $type, null otherwise
+     */
+    public static function map($type = null, $className = null)
+    {
+        if ($type === null) {
+            return self::$_types;
+        }
+        if (!is_string($type)) {
+            self::$_types = $type;
+            return null;
+        }
+        if ($className === null) {
+            return isset(self::$_types[$type]) ? self::$_types[$type] : null;
+        }
+        self::$_types[$type] = $className;
+    }
+
+    /**
+     * Clears out all created instances and mapped types classes, useful for testing
+     *
+     * @return void
+     */
+    public static function clear()
+    {
+        self::$_types = [];
+        self::$_builtTypes = [];
+    }
+
+    /**
+     * Casts given value from a PHP type to one acceptable by database
+     *
+     * @param mixed $value value to be converted to database equivalent
+     * @param Driver $driver object from which database preferences and configuration will be extracted
+     * @return mixed
+     */
+    public function toDatabase($value, Driver $driver)
```

Comment:
I don't see why the `toDatabase` method couldn't be a simple alias for `toDatasource`?  I don't think its strange for the database API to use a method with 'datasource' in it.